### PR TITLE
New webhook for adding dco-check webhook to newly created/publicized repos

### PR DESCRIPTION
**Pull request by robbkidd**
Depends on the changes in #6 for the relocation of the gem files into `src/` because cleanliness is next to and all that.

### CODE REVIEW

**Comment by robbkidd** [comment]
For you, @tas50. I plan to spin a test instance of this up and point a personal play-toy organization at it.

File: src/lib/dcob.rb [hunk 1]
```diff
@@ -86,6 +104,11 @@ def verify_signature(payload_body)
     puts "You must specify cfg.dco_info_url in config.toml"
     exit 1
   end
-end
 
-Dcob::Server.run!
+  if config["cfg"]["dco_webhook_url"]
+    DCO_WEBHOOK_URL = config["cfg"]["dco_webhook_url"]
+  else
+    puts "You must specify cfg.dco_webhook_url in config.toml"
```
Comment (robbkidd) [inline src/lib/dcob.rb]: I wonder if this really needs to be configurable. The webhook URL (e.g. "me" from the bot's perspective) could be determined from the incoming `request.url` or `request.scheme + '://' + request.host + '/payload'`

File: src/lib/dcob.rb [hunk 2]
```diff
@@ -1,49 +1,66 @@
...
+    end
+
+    set(:event_type) do |type|
+      condition { request.env["X-GitHub-Event"] == type }
+    end
+
+    post "/payload", event_type: "repository" do
+      begin
+        repo = JSON.parse(@payload_body)
+        if %w{created publicized}.include? repo["action"]
```
Comment (robbkidd) [inline src/lib/dcob.rb]: Also filter on only public repos.

File: habitat/hooks/run [hunk 3]
```diff
@@ -0,0 +1,14 @@
+#!/bin/sh
+cd /hab/svc/dcob
+
+export RACK_ENV=production
+export SSL_CERT_FILE="$(hab pkg path core/cacerts)/ssl/certs/cacert.pem"
```
Comment (robbkidd) [inline habitat/hooks/run]: Makes me weepy to do this now. I suppose if there is a way to write out our own run hook in some phase of the plan, we could generate this at build time.

**Comment by robbkidd** [comment]
Got the bot working on my test org. I'm really to review whenever.

File: habitat/hooks/run (hunk 3 above)
Comment (robbkidd) [reply habitat/hooks/run]: Gonna roll this back to a `pkg_svc_run` command in the plan and make setting these a requirement prior to running `hab start chef/dcob`.

File: src/lib/dcob/octoclient.rb [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+    end
+
+    def hookit(repository, callback_url)
+      github_service_name = "web"
+      hook_config = { url: callback_url,
+                      content_type: "json",
+                      secret: SECRET_TOKEN }
+      hook_options = { events: ["pull_request"],
+                       active: true }
+      Octokit.create_hook(repository, github_service_name, hook_config, hook_options)
```
Comment (robbkidd) [inline src/lib/dcob/octoclient.rb]: Saw a timing problem with the create_hook being possibly being attempted before the repository was fully created at GitHub causing this to fail. Add a retry loop. 😢

**Comment by robbkidd** [comment]
Created a few repositories in the test org with the version of the bot that includes the retry loop. All successfully received the repo webhook. I don't know whether retries were involved, though, because I haven't sorted out bot logging yet. I supposed that might be next.

**Review by thommay — APPROVED** [review]
Once that hook is gone I'm 👍

**Review by thommay — APPROVED** [review]

**Pull request merged**